### PR TITLE
MAINT/TST: Fix/Silence DeprecationWarnings

### PR DESCRIPTION
**Pull request by jbrockmendel**
Each commit addresses a single source of these warnings.

Kept peripheral cleanup to a near minimum (the case that strayed farthest from this goal was that last commit, which touches regression._prediction and genmod._prediction).

It's conceivable that the way _penalties currently works is intentional, but it looks like `grad` was deprecated but internal usage was not updated to reflect that.  If this perception is wrong then this commit will have to be reverted and those warnings addressed separately.

Made an executive decision to stop passing through numpy's `RuntimeWarning`s when `tvalues` includes a division by zero.  On balance those tend to occur when other convergence-type problems occur, and an additional warning isn't all that helpful.

`patsy` change to make `design_info.builder` not need the `.builder` is effective as of patsy4.0, which is the required version according to setup.py

### CODE REVIEW

File: statsmodels/base/data.py [hunk 1]
```diff
@@ -475,14 +474,16 @@ def _convert_endog_exog(self, endog, exog=None):
 
     @classmethod
     def _drop_nans(cls, x, nan_mask):
-        if hasattr(x, 'ix'):
+        if hasattr(x, 'loc'):
```
Comment (bashtage) [inline statsmodels/base/data.py]: cleaner to check for Series, DataFrame

File: statsmodels/base/data.py [hunk 2]
```diff
@@ -475,14 +474,16 @@ def _convert_endog_exog(self, endog, exog=None):
...
+            # i.e. is pandas Series/DataFrame
             return x.loc[nan_mask]
         else:  # extra arguments could be plain ndarrays
             return super(PandasData, cls)._drop_nans(x, nan_mask)
 
     @classmethod
     def _drop_nans_2d(cls, x, nan_mask):
-        if hasattr(x, 'ix'):
+        if hasattr(x, 'loc'):
```
Comment (bashtage) [inline statsmodels/base/data.py]: same

File: statsmodels/base/model.py [hunk 3]
```diff
@@ -1257,7 +1257,10 @@ def tvalues(self):
         """
         Return the t-statistic for a given parameter estimate.
         """
-        return self.params / self.bse
+        with np.errstate(divide="ignore"):
```
Comment (bashtage) [inline statsmodels/base/model.py]: Should be shown to end users.  Probably don't suppress here or in tests.

File: statsmodels/nonparametric/kernel_regression.py [hunk 4]
```diff
@@ -767,9 +767,9 @@ def _compute_se_lambda(self, Y, X):
         but slows down computation significantly.
         """
         n = np.shape(Y)[0]
-        lam = np.empty(shape=(self.nres, ))
+        lam = np.empty(shape=(self.nres,))
         for i in range(self.nres):
-            ind = np.random.random_integers(0, n-1, size=(n,1))
+            ind = np.random.randint(0, n-1, size=(n, 1))
```
Comment (bashtage) [inline statsmodels/nonparametric/kernel_regression.py]: Needs to be (0,n,size=(n,1)) -> not the deprecation warning saying +1. Former was closed.  This is open.

File: statsmodels/nonparametric/kernel_regression.py [hunk 5]
```diff
@@ -793,12 +793,12 @@ def _compute_sig(self):
...
         # Calculate the restricted mean. See p. 372 in [8]
         M = KernelReg(Y, X, self.var_type, self.model.reg_type, self.bw,
-                      defaults = EstimatorSettings(efficient=False)).fit()[0]
+                      defaults=EstimatorSettings(efficient=False)).fit()[0]
         M = np.reshape(M, (n, 1))
         e = Y - M
         e = e - np.mean(e)  # recenter residuals
         for i in range(self.nboot):
-            ind = np.random.random_integers(0, n-1, size=(n,1))
+            ind = np.random.randint(0, n-1, size=(n, 1))
```
Comment (bashtage) [inline statsmodels/nonparametric/kernel_regression.py]: Same, (0,n)

File: statsmodels/sandbox/nonparametric/kernel_extras.py [hunk 6]
```diff
@@ -75,7 +75,7 @@ class TestFForm(object):
     References
     ----------
     See Racine, J.: "Consistent Significance Testing for Nonparametric
-    Regression" Journal of Business \& Economics Statistics.
+    Regression" Journal of Business \\& Economics Statistics.
```
Comment (bashtage) [inline statsmodels/sandbox/nonparametric/kernel_extras.py]: Why does this need any escaping?

File: statsmodels/sandbox/regression/penalized.py [hunk 7]
```diff
@@ -59,8 +59,8 @@ class TheilGLS(GLS):
 
     .. math:: q = R \beta + v
```
Comment (bashtage) [inline statsmodels/sandbox/regression/penalized.py]: Better to make the docstring raw and leave single escapes than to mix and match escaped-escapes and non-escape sequences (e.g., \\ and \b)

File: statsmodels/sandbox/regression/penalized.py [hunk 8]
```diff
@@ -59,8 +59,8 @@ class TheilGLS(GLS):
...
 
-    :math:`E(u) = 0`, :math:`cov(u) = \Sigma`
-    :math:`cov(u, v) = \Sigma_p`, with full rank.
+    :math:`E(u) = 0`, :math:`cov(u) = \\Sigma`
+    :math:`cov(u, v) = \\Sigma_p`, with full rank.
```
Comment (bashtage) [inline statsmodels/sandbox/regression/penalized.py]: raw docstrings are just better with LaTeX math

**Comment by bashtage** [comment]
I have quite a few that you don't have. but we have overlap, so will have conflicts  Feel free to take them from mine into this and I'll close mine.

**Comment by bashtage** [comment]
I should add the dead simplest way to avoid these warnings (temporarily) is to 

pip install pytest==3.7

which will pin to the last version that worked.

**Comment by jbrockmendel** [comment]
> Only the massive set of new warnings should be suppressed. 0 div should still be allowed since these are really errors in test design.

Fair enough.

> so will have conflicts Feel free to take them from mine into this and I'll close mine.

I was thinking that you're more likely to get this past the maintainer in a timely manner than I am, so we should go the other way.

File: statsmodels/base/model.py (hunk 3 above)
Comment (jbrockmendel) [reply statsmodels/base/model.py]: Will revert here; merits discussion elsewhere.

**Comment by josef-pkt** [comment]
Can your make a PR with just this:
`pip install pytest==3.7`

I'm not sure whether there is not also a bug in pytest 3.8, and I would rather not change a lot of things right away, and let other packages figure this out first.
I'm a bit suspicious whether all this backslashes are really incorrect.

File: statsmodels/sandbox/nonparametric/kernel_extras.py (hunk 6 above)
Comment (jbrockmendel) [reply statsmodels/sandbox/nonparametric/kernel_extras.py]: I'd be happy without; assumed previous author had a reason.  Will change.

File: statsmodels/sandbox/regression/penalized.py (hunk 7 above)
Comment (jbrockmendel) [reply statsmodels/sandbox/regression/penalized.py]: Yah, wasn't super-happy with this mixed case thing.

**Comment by jbrockmendel** [comment]
> I'm not sure whether there is not also a bug in pytest 3.8

What bug are you referring to?

> I'm a bit suspicious whether all this backslashes are really incorrect.

Referring to regexes or docstrings or both?

> and I would rather not change a lot of things 

Totally out of character.

bashtage and I have unfortunate timing when it comes to trying to be helpful.  I think the thing to do is merge #5135 when feasible and I'll rebase+repush with just the non-overlapping changes.

**Comment by bashtage** [comment]
All of the back slashes I looked at were incorrect. Mostly missing raw
string flags.

**Comment by josef-pkt** [comment]
what about the regexes like `\[` ?
If you switch back to older pytest, then we see whether the unit tests actually still pass, instead of trying to get results through the warnings avalanche.

(Even if everything is fine, then I don't want to see the same deprecation warnings a thousand times each time pandas or numpy changes something that affects patsy or statsmodels.)

**Comment by jbrockmendel** [comment]
Rebased following #5135.

**Comment by bashtage** [comment]
LGTM

**Comment by jbrockmendel** [comment]
@josef-pkt should I bother rebasing or is that just going to clog the CI?

**Comment by bashtage** [comment]
Should be rebased and merged.  Solves, among other issues, statsmodels calling code that statsmodels has deprecated.

**Comment by jbrockmendel** [comment]
> Should be rebased and merged. Solves, among other issues, statsmodels calling code that statsmodels has deprecated.

I very much agree.  But my understanding is that my PRs are currently being actively ignored, so until that is over pushing will only serve to clog the CI.

**Comment by jbrockmendel** [comment]
@josef-pkt this actually includes a non-trivial thing that does need your attention

**Pull request closed without merging**